src/dhcp/clientv4.rs

```diff
@@ -0,0 +1,458 @@
+                               self.dns_servers = dns_servers;
+                           }
+                   }
+                   None => {}
+               }
+        }
+
+        match self.state {
+            ClientState::Discovering
+                if dhcp_repr.message_type == DhcpMessageType::Offer =>
+            {
+                self.next_egress = now;
+                let r_state = RequestState {
+                    retry: 0,
+                    endpoint_ip: src_ip,
+                    server_identifier,
+                };
+                Some(ClientState::Requesting(r_state))
+            }
+            ClientState::Requesting(ref r_state)
+                if dhcp_repr.message_type == DhcpMessageType::Ack &&
+                   server_identifier == r_state.server_identifier =>
+            {
+                self.next_egress = now + Duration::from_secs(RENEW_INTERVAL);
+                let p_state = RenewState {
+                    retry: 0,
+                    endpoint_ip: src_ip,
+                    server_identifier,
+                };
+                Some(ClientState::Renew(p_state))
+            }
+            ClientState::Renew(ref mut p_state)
+                if dhcp_repr.message_type == DhcpMessageType::Ack &&
+                server_identifier == p_state.server_identifier =>
+            {
+                self.next_egress = now + Duration::from_secs(RENEW_INTERVAL);
+                p_state.retry = 0;
+                None
+            }
+            _ => None
+        }.map(|new_state| self.state = new_state);
+    }
+
+    fn egress<DeviceT: for<'d> Device<'d>>(&mut self, iface: &mut Interface<DeviceT>, raw_socket: &mut RawSocket, checksum_caps: &ChecksumCapabilities, now: Instant) -> Result<()> {
+        self.transaction_id += 1;
+        let mac = iface.ethernet_addr();
+
+        match self.state {
+            ClientState::Discovering => {
+                self.next_egress = now + Duration::from_secs(DISCOVER_TIMEOUT);
+
+                let endpoint = IpEndpoint {
+                    addr: Ipv4Address::BROADCAST.into(),
+                    port: UDP_SERVER_PORT,
+                };
+                let dhcp_repr = DhcpRepr {
+                    message_type: DhcpMessageType::Discover,
+                    transaction_id: self.transaction_id,
+                    client_hardware_address: mac,
+                    client_ip: Ipv4Address::UNSPECIFIED,
+                    your_ip: Ipv4Address::UNSPECIFIED,
+                    server_ip: Ipv4Address::UNSPECIFIED,
+                    router: None,
+                    subnet_mask: None,
+                    relay_agent_ip: Ipv4Address::UNSPECIFIED,
+                    broadcast: true,
+                    requested_ip: None,
+                    client_identifier: Some(mac),
+                    server_identifier: None,
+                    parameter_request_list: None,
+                    dns_servers: None,
```

Comment:
Can you factor out the common fields here, so that they're only filled once at the beginning of the function?